### PR TITLE
Ubuntu 22.04 で Intel VPL の H.265 に対応する

### PR DESCRIPTION
**Pull request by enm10k**
## 変更内容

- Ubuntu 22.04 で Intel VPL の H.265 に対応しました

## メモ

libvpl の更新に伴い、 Intel の apt リポジトリから入手したドライバーを利用しないと一部機能が利用できないことを確認しています

```
# Ubuntu 22.04 にドライバーをインストールする手順

## apt リポジトリの追加
wget -qO - https://repositories.intel.com/gpu/intel-graphics.key | \
  sudo gpg --dearmor --output /usr/share/keyrings/intel-graphics.gpg
echo "deb [arch=amd64,i386 signed-by=/usr/share/keyrings/intel-graphics.gpg] https://repositories.intel.com/gpu/ubuntu jammy client" | \
  sudo tee /etc/apt/sources.list.d/intel-gpu-jammy.list
sudo apt update

## インストール
sudo apt install -y \
  intel-opencl-icd intel-level-zero-gpu level-zero \
  intel-media-va-driver-non-free libmfx1 libmfx-gen1.2 libvpl2 \
  libegl-mesa0 libegl1-mesa libegl1-mesa-dev libgbm1 libgl1-mesa-dev libgl1-mesa-dri \
  libglapi-mesa libgles2-mesa-dev libglx-mesa0 libigdgmm12 libxatracker2 mesa-va-drivers \
  mesa-vdpau-drivers mesa-vulkan-drivers va-driver-all vainfo hwinfo clinfo
```

参考: https://dgpu-docs.intel.com/driver/client/overview.html

---

This pull request includes changes to the `.vscode/c_cpp_properties.json` file, `CHANGES.md`, `CMakeLists.txt`, `VERSION`, `buildbase.py`, `examples/VERSION`, `default_video_formats.cpp`, and several files in the `src` directory. The changes mainly involve updating the version of `libwebrtc` and `libvpl`, adding support for Ubuntu 22.04, and adding support for the H.265 codec.

Updates to libraries and system support:

* `VERSION`, `examples/VERSION`: Updated the `WEBRTC_BUILD_VERSION` to `m122.6261.1.0` and `VPL_VERSION` to `v2.10.1`. [[1]](diffhunk://#diff-7b60b8e351cbb80c47459ffe2c79f1a26404871f49294780fe47ad0e58c09350L2-R9) [[2]](diffhunk://#diff-03c0d8fccd99f23bd7d8ebbf740a199bbf211cc71e7deffec280e236f7026a3bL2-R2)
* [`CHANGES.md`](diffhunk://#diff-d975bf659606195d2165918f93e1cf680ef68ea3c9cab994f033705fea8238b2R16-R26): Documented the updates to `libwebrtc` and `libvpl`, and added support for Ubuntu 22.04 and Intel VPL's H.265 codec.
* [`.vscode/c_cpp_properties.json`](diffhunk://#diff-8f4e8cf66ff6479868eeeb084ef19fbb8bea6102cf86f8adec2180dcf38dc47dL29-R29): Added a new configuration for Ubuntu 22.04 and modified the `USE_NVCODEC_ENCODER` and `USE_JETSON_ENCODER` defines. [[1]](diffhunk://#diff-8f4e8cf66ff6479868eeeb084ef19fbb8bea6102cf86f8adec2180dcf38dc47dL29-R29) [[2]](diffhunk://#diff-8f4e8cf66ff6479868eeeb084ef19fbb8bea6102cf86f8adec2180dcf38dc47dL59-R59) [[3]](diffhunk://#diff-8f4e8cf66ff6479868eeeb084ef19fbb8bea6102cf86f8adec2180dcf38dc47dR69-R102)
* [`CMakeLists.txt`](diffhunk://#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR250): Added `RTC_ENABLE_H265` to the compile definitions for Windows and Ubuntu. [[1]](diffhunk://#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR250) [[2]](diffhunk://#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR423)

Changes to H.265 codec support:

* [`default_video_formats.cpp`](diffhunk://#diff-bb219ab76a043fcfe343ebb33875c8f056683b67b0f98a67d0d8b971088da1e8R39-R40): Added a new condition for the `webrtc::kVideoCodecH265` codec.
* `vpl_video_decoder.cpp`, `vpl_video_encoder.cpp`: Added several changes to support the H.265 codec, including updating the `VplVideoEncoderImpl` and `VplVideoDecoderImpl` classes and modifying the `CreateEncoder` function. [[1]](diffhunk://#diff-53ff2f81fc7f325a26985b65f0e4d2283c7a0dc63b976e646e001c93ce8cecbaR131-R134) [[2]](diffhunk://#diff-53ff2f81fc7f325a26985b65f0e4d2283c7a0dc63b976e646e001c93ce8cecbaL144-R165) [[3]](diffhunk://#diff-53ff2f81fc7f325a26985b65f0e4d2283c7a0dc63b976e646e001c93ce8cecbaL278-R279) [[4]](diffhunk://#diff-394c52d594b3fc36acc45e6f42f0d974d445ef2b0cb5e174246756a80037f39aL4-R11) [[5]](diffhunk://#diff-394c52d594b3fc36acc45e6f42f0d974d445ef2b0cb5e174246756a80037f39aR69) [[6]](diffhunk://#diff-394c52d594b3fc36acc45e6f42f0d974d445ef2b0cb5e174246756a80037f39aR121-L129) [[7]](diffhunk://#diff-394c52d594b3fc36acc45e6f42f0d974d445ef2b0cb5e174246756a80037f39aR182-R192) [[8]](diffhunk://#diff-394c52d594b3fc36acc45e6f42f0d974d445ef2b0cb5e174246756a80037f39aR203-R210)

Changes to logging and error handling:

* [`vpl_session_impl.cpp`](diffhunk://#diff-6ce6b6da1c5258033cd7f484c2c6d46cd569a02952b75a9e46b0b17392e1b87aL3-R3): Replaced `std::cerr` with `RTC_LOG(LS_VERBOSE)` for logging and error handling. [[1]](diffhunk://#diff-6ce6b6da1c5258033cd7f484c2c6d46cd569a02952b75a9e46b0b17392e1b87aL3-R3) [[2]](diffhunk://#diff-6ce6b6da1c5258033cd7f484c2c6d46cd569a02952b75a9e46b0b17392e1b87aL38-R38) [[3]](diffhunk://#diff-6ce6b6da1c5258033cd7f484c2c6d46cd569a02952b75a9e46b0b17392e1b87aL47-R47) [[4]](diffhunk://#diff-6ce6b6da1c5258033cd7f484c2c6d46cd569a02952b75a9e46b0b17392e1b87aL69-R82)
* `vpl_video_decoder.cpp`, `vpl_video_encoder.cpp`: Updated logging and error handling, including adding verbose logging and updating the handling of unsupported codecs. [[1]](diffhunk://#diff-53ff2f81fc7f325a26985b65f0e4d2283c7a0dc63b976e646e001c93ce8cecbaL144-R165) Fde3f2e4L179R179, [[2]](diffhunk://#diff-394c52d594b3fc36acc45e6f42f0d974d445ef2b0cb5e174246756a80037f39aR203-R210)

Changes to source code repositories:

* [`buildbase.py`](diffhunk://#diff-2e3bbef7102f8e50528a74a67436d9a0c971c85f1479ac9b86013afdd57dc59aL1066-R1066): Changed the source repository for `libvpl` from `oneapi-src/oneVPL.git` to `intel/libvpl.git`.

### CODE REVIEW

File: src/hwenc_vpl/vpl_video_encoder.cpp [hunk 1]
```diff
@@ -463,15 +516,22 @@ int32_t VplVideoEncoderImpl::Encode(
       encoded_image_._frameType = webrtc::VideoFrameType::kVideoFrameDelta;
     }
 
+    if (codec_ == MFX_CODEC_AVC) {
+      h264_bitstream_parser_.ParseBitstream(encoded_image_);
+      encoded_image_.qp_ = h264_bitstream_parser_.GetLastSliceQp().value_or(-1);
+    }
```
Comment (enm10k) [inline src/hwenc_vpl/vpl_video_encoder.cpp:522]: @melpon 
libwebrtc のコードを確認したところ、 M122 の時点で H265BitstreamParser クラスが存在しており、 H.264 と同様に H.265 でも  `encoded_image.qp_` に QP を設定するコードが書けそうです

こちらにもそのような処理を追加すべきでしょうか?

参考: https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/common_video/h265/h265_bitstream_parser.cc

Comment (melpon) [reply src/hwenc_vpl/vpl_video_encoder.cpp:522]: やった方が良いと思います

**Comment by melpon** [comment]
ちょっと全体的に、自分が雑にやった部分が戻されて無くて、無駄なログ、間違った種別のログ、不要なコメントとか残ったままなので、一通り見直して下さい。

**Comment by enm10k** [comment]
@melpon 

> ちょっと全体的に、自分が雑にやった部分が戻されて無くて、無駄なログ、間違った種別のログ、不要なコメントとか残ったままなので、一通り見直して下さい。

以下のようにしようと思いますが問題ないでしょうか?
認識に相違があれば指摘してください

- 不要なコメント
  - H265 以外のコーデックでもパラメーターを設定する行でコメントアウトされているものがそこそこありますが、基本的に使われていない行は削除します
  - パラメーターをデバッグする目的のマクロは残します
- ログ
  - エンコーダーの初期化が失敗した際のログは極力 LS_WARNING / LS_ERROR で残します
  - エンコード時のログについては、不要そうなものを消します
    - `Key frame {requested:generated}` など

**Comment by melpon** [comment]
> エンコーダーの初期化が失敗した際のログは極力 LS_WARNING / LS_ERROR で残します

これに関しては、コメントアウトするか、IsSupported 経由で呼び出したときはログに出力しないようにするかですね。
IsSupported でエンコード能力を確認してる時にエラーが出まくるのは良くないです。

**Comment by enm10k** [comment]
@melpon 
ありがとうございます

> これに関しては、コメントアウトするか、IsSupported 経由で呼び出したときはログに出力しないようにするかですね。
> IsSupported でエンコード能力を確認してる時にエラーが出まくるのは良くないです。

仰る通り、エンコード能力を確認している時にエラーが出まくるのは良くないですね
一方で、サポート/デバッグ目的としては、一部のログは残したい気もします

LS_ERROR や LS_WARNING ではなく、 LS_VERBOSE でログを残すのはダメでしょうか?

**Comment by melpon** [comment]
LS_DEBUG 以下なら良さそうな気はしますね

**Comment by enm10k** [comment]
@melpon 修正したので、再度ご確認いただきたいです

> LS_DEBUG 以下なら良さそうな気はしますね

LS_DEBUG が存在しなかったため、 LS_VERBOSE にしました
参照: https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/rtc_base/logging.h;l=82-97;drc=d212551ad3be466a45c4e428f11c86c53b962887

File: src/hwenc_vpl/vpl_video_encoder.cpp [hunk 2]
```diff
@@ -539,7 +557,7 @@ int32_t VplVideoEncoderImpl::InitVpl() {
   VPL_CHECK_RESULT(sts, MFX_ERR_NONE, sts);
 
   RTC_LOG(LS_INFO) << "Encoder NumFrameSuggested="
-                   << alloc_request_.NumFrameSuggested;
+                    << alloc_request_.NumFrameSuggested;
```
Comment (melpon) [inline src/hwenc_vpl/vpl_video_encoder.cpp:560]: ここ１個空白の差分あるのなんだろう。以前のやつの方が `<<` が揃ってて、ちゃんとフォーマッタ掛けたら以前のやつに戻りそうな感じがします。

File: src/hwenc_vpl/vpl_video_encoder.cpp [hunk 3]
```diff
@@ -179,6 +179,17 @@ std::unique_ptr<MFXVideoENCODE> VplVideoEncoderImpl::CreateEncoder(
     ext_buffers[0] = (mfxExtBuffer*)&ext_coding_option;
     ext_buffers[1] = (mfxExtBuffer*)&ext_coding_option2;
     ext_buffers_size = 2;
+  } else if (codec == MFX_CODEC_HEVC) {
+    // InitialDelayInKB を MFX_CODEC_AVC で設定するとエラーになる
+    param.mfx.InitialDelayInKB = target_kbps;
```
Comment (melpon) [inline src/hwenc_vpl/vpl_video_encoder.cpp:184]: これ無くても動くなら無くしちゃって良いと思います。

File: src/hwenc_vpl/vpl_video_decoder.cpp [hunk 4]
```diff
@@ -128,6 +128,10 @@ std::unique_ptr<MFXVideoDECODE> VplVideoDecoderImpl::CreateDecoderInternal(
   param.AsyncDepth = 1;
   param.IOPattern = MFX_IOPATTERN_OUT_SYSTEM_MEMORY;
 
+  if (codec == MFX_CODEC_HEVC) {
+    param.mfx.CodecProfile = MFX_PROFILE_HEVC_MAIN;
+  }
```
Comment (melpon) [inline src/hwenc_vpl/vpl_video_decoder.cpp:133]: 流れ的に、この処理は CodecId をセットした直後の方がいいかな

**Review by melpon — APPROVED** [review]

File: src/hwenc_vpl/vpl_video_decoder.cpp [hunk 5]
```diff
@@ -275,7 +276,7 @@ int32_t VplVideoDecoderImpl::Decode(const webrtc::EncodedImage& input_image,
   // H264 は sts == MFX_WRN_VIDEO_PARAM_CHANGED でハンドリングできるのでここではチェックしない
   // VP9 は受信フレームのサイズが変わっても MFX_WRN_VIDEO_PARAM_CHANGED を返さないようなので、
   // ここで毎フレーム情報を取得してサイズを更新する。
-  if (codec_ != MFX_CODEC_AVC) {
+  if (codec_ != MFX_CODEC_AVC || codec_ != MFX_CODEC_HEVC) {
```
[the remark this answers is not in the document]
Comment (enm10k) [reply src/hwenc_vpl/vpl_video_decoder.cpp:279]: ありがとうございます
この際 `codec_ == MFX_CODEC_VP9` 的な条件にしてはどうかと思いましたがいかがですか?
(AV1 の場合もこの分岐に入るべきでしょうか?)

[the remark this answers is not in the document]
Comment (melpon) [reply src/hwenc_vpl/vpl_video_decoder.cpp:279]: 確かになんでこんな条件にしたんだろ…。コメント見る限り `== MFX_CODEC_VP9` で良さそうなので、そうしちゃって下さい。

File: src/hwenc_vpl/vpl_video_encoder.cpp (hunk 3 above)
Comment (enm10k) [reply src/hwenc_vpl/vpl_video_encoder.cpp]: 動いたので [a7aa584](https://github.com/shiguredo/sora-cpp-sdk/pull/82/commits/a7aa584a857e7faf801a410e42546e21a9dc9a94) で消しました

File: buildbase.py [hunk 6]
```diff
@@ -1063,7 +1063,7 @@ def install_vpl(version, configuration, source_dir, build_dir, install_dir, cmak
     rm_rf(vpl_source_dir)
     rm_rf(vpl_build_dir)
     rm_rf(vpl_install_dir)
-    git_clone_shallow("https://github.com/oneapi-src/oneVPL.git", version, vpl_source_dir)
+    git_clone_shallow("https://github.com/intel/libvpl.git", version, vpl_source_dir)
```
Comment (torikizi) [inline buildbase.py:1066]: oneapi-src から取得していたところを intel の libvpl から取得するように変更になっていますが、これまで動作していた Media SDK の環境では動作するのでしょうか。
この点何かあるのでしたら CHANGES に補足があると良さそうです。

**Review by torikizi — APPROVED** [review]

File: buildbase.py (hunk 6 above)
Comment (enm10k) [reply buildbase.py:1066]: こちらは GitHub リポジトリ自体がリダイレクトされていた & リダイレクト先のリポジトリにも古いタグが残っているので、リポジトリを書き換えたことによる Media SDK への影響はないはずです

libvpl (oneVPL) のバージョンをあげているので、それが原因で問題が起きる可能性はあります

Comment (torikizi) [reply buildbase.py:1066]: 承知しました。
ありがとうございます！

File: src/hwenc_vpl/vpl_video_decoder.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (enm10k) [reply src/hwenc_vpl/vpl_video_decoder.cpp:281]: わかりました > `== MFX_CODEC_VP9` で良さそう

ちなみに `MFX_WRN_VIDEO_PARAM_CHANGED` がこの条件分岐の少し前にハンドルされており、 width_, height_ が更新された場合は continue で抜けてここまで来ないので、コーデックの分岐をなくすことも可能かもしれませんね
(その場合、 VP9 以外のコーデックでも GetVideoParam が毎回実行されるようになるため、処理のコスト的に問題ないか? の確認は必要そうです)

**Review by melpon — APPROVED** [review]
よさそう

**Comment by enm10k** [comment]
ありがとうございます。
マージします。

**Pull request merged**